### PR TITLE
Rewrite to fit CBA template

### PR DESCRIPTION
**Pull request by johnb432**
I really like the concept of this mod, but I felt like it could do with some changes and optimizations:
- Uses the CBA Template for making PBOs.
- Added support for making the mod using [HEMTT](https://github.com/BrettMayson/HEMTT).
- Added support for [EditorConfig](https://editorconfig.org/).
- Version of the mod can be defined inside `script_version.hpp`
- Usage of CBA macros.
- Usage of `CBA_fnc_addPerFrameHandler` for running the loop.
- Optimized networking when it comes to resetting a unit's camo coefficient.
- Changed two CBA settings to allow more customization (ghillie reduction and night compensation).

This PR still needs:
- `script_version.hpp` to be updated
- Some testing: E.g. I have checked if `fnc_sendDebugInfo` works as intended.
- There is some code that tries to determine the ambient light and see off of that if it's night time or not. This still needs some polishing.

### CODE REVIEW

**Comment by rekterakathom** [comment]
Thanks a lot for this! This is something that I wanted to do back in July but didn't have the time for. Looks good to me, though I will run a lot of tests to ensure that this works as it should before merging and updating the workshop version.

File: addons/core/initSettings.sqf [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// ADVANCED SETTINGS
```
Comment (rekterakathom) [inline addons/core/initSettings.sqf:1]: ``#include "script_macros.hpp"`` here? Currently this won't compile with ArmaScriptCompiler

**Comment by rekterakathom** [comment]
Camouflage coefficient isn't resetting to 1 when disabling the system from CBA settings. Tested with a single unit on Stratis.

**Comment by rekterakathom** [comment]
Not really an issue but I'd like to know regardless: why are units stored hashed in ``unitCache``? For security reasons I'd presume?

**Comment by johnb432** [comment]
> Camouflage coefficient isn't resetting to 1 when disabling the system from CBA settings. Tested with a single unit on Stratis.

Should be fixed now.


> Not really an issue but I'd like to know regardless: why are units stored hashed in `unitCache`? For security reasons I'd presume? I think `XEH_postInit.sqf` could contain some comments that briefly explain the caches?

The unit cache exists to avoid sending the same calculated coefficient over the network multiple times.

File: addons/core/initSettings.sqf (hunk 1 above)
Comment (johnb432) [reply addons/core/initSettings.sqf:3]: Should be resolved.

File: addons/core/functions/fnc_loop.sqf [hunk 2]
```diff
@@ -22,31 +22,42 @@ if (!isNil QGVAR(PFHID)) then {
...
+        GVAR(PFHID) = nil;
     };
 
     // Get all affected units
     private _targets = switch (CBA_SETTING(targetList)) do {
-        case 0: {call CBA_fnc_players};
+        case 2: {allUnits};
         case 1: {flatten ((allGroups select {isPlayer leader _x}) apply {units _x})};
-        default {allUnits};
+        default {call CBA_fnc_players};
```
Comment (rekterakathom) [inline addons/core/functions/fnc_loop.sqf:57]: Do we want ``CBA_fnc_players`` here? From my tests it is a lot slower than ``allPlayers - entities "HeadlessClient_F"`` and IIRC only needed when the mission hasn't been properly initialized yet

Comment (johnb432) [reply addons/core/functions/fnc_loop.sqf:57]: I don't think we absolutely need `CBA_fnc_players`, as you pointed out. I will change it to your suggestion.

**Comment by rekterakathom** [comment]
I don't think HEMTT is able to produce releases with the current project structure, as it doesn't find ``script_version.hpp`` since the expected path is ``addons/main/script_version.hpp``. In my opinion this is not a big deal and simply adding the version to ``hemtt.toml``

File: .hemtt/base.toml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+name = "Dynamic-Camo-System"
```
Comment (rekterakathom) [inline .hemtt/base.toml:1]: Add ``modname = "Dynamic Camo System"`` so the name of the addon stays consistent with what it was before. If the name of an addon changes and it's published to the workshop, Steam will download the addon again under the new name, essentially creating a duplicate.

**Comment by rekterakathom** [comment]
I will update the mod to the workshop with this PR, as version 2.0.0.0, due to the remaining changes not being significant in any way to the function of the mod. The remaining small nitpicks can be ironed out and the PR merged later.

**Comment by johnb432** [comment]
Sorry for the late response.
I have pushed both suggestions made above. However, I can't check if they work properly as I still use HEMTT 0.7.6

**Pull request merged**